adlfs/transfer.py

```diff
@@ -267,110 +310,101 @@ def progress(self):
-                        dic['merge'] = self._submit(self._merge, self._adlfs,
-                                                    dst, chunks)
-                    else:
-                        dic['stop'] = time.time()
-                        self._fstates[(src, dst)] = 'finished'
-                        self._status(src, dst, dic['nbytes'], dic['start'], dic['stop'])
-                elif dic['cstates'].contains_none('running'):
-                    logger.debug("Transfer failed: %s", dic['cstates'])
-                    self._fstates[(src, dst)] = 'errored'
-                else:
-                    logger.debug("Transferring chunks: %s", dic['cstates'])
-            elif self._fstates[(src, dst)] == 'merging':
-                future = dic['merge']
-                if not future.done():
-                    continue
-                if future.cancelled():
-                    self._fstates[(src, dst)] = 'cancelled'
-                elif future.exception():
-                    self._fstates[(src, dst)] = 'errored'
+    def _update(self, future):
+        if future in self._cfutures:
+            obj = self._cfutures[future]
+            parent = self._chunks[obj]['parent']
+            cstates = self._files[parent]['cstates']
+
+            if future.cancelled():
+                cstates[obj] = 'cancelled'
+            elif future.exception():
+                cstates[obj] = 'errored'
+            else:
+                self._chunks[obj]['actual'] = future.result()
+                cstates[obj] = 'finished'
+
+            if cstates.contains_all('finished'):
+                logger.debug("Chunks transferred")
+                src, dst = parent
+                if self._merge and len(cstates.objects) > 1:
+                    logger.debug("Merging file: %s", self._fstates[parent])
+                    self._fstates[parent] = 'merging'
+                    merge_future = self._submit(
+                        self._merge, self._adlfs, dst,
+                        [o[0] for o in cstates.objects])
+                    self._ffutures[merge_future] = parent
                 else:
-                    dic['stop'] = time.time()
-                    self._fstates[(src, dst)] = 'finished'
-                    self._status(src, dst, dic['nbytes'], dic['start'], dic['stop'])
+                    self._fstates[parent] = 'finished'
+                    self._files[parent]['stop'] = time.time()
+                    self._status(src, dst)
+            elif cstates.contains_none('running'):
+                logger.debug("Transfer failed: %s", cstates)
+                self._fstates[parent] = 'errored'
+        elif future in self._ffutures:
+            src, dst = self._ffutures[future]
+
+            if future.cancelled():
+                self._fstates[(src, dst)] = 'cancelled'
+            elif future.exception():
+                self._fstates[(src, dst)] = 'errored'
+            else:
+                result = future.result()
```

Comment:
Should we not check the number of bytes here?
Also, here is where we would check for False (or the string repr of the exception, perhaps) in case of exception caught in the thread and not passed back.